#QUBOで素因数分解
ここでは、QUBOを活用した素因数分解の方法を見ていきます。素因数分解は最小値問題に落とし込むことができ、コスト関数のコストを見ることで正解にたどり着いているかどうかを確認することができます。

また、素因数分解は膨大な解候補の中からたった１つの解を求めないといけないのでQUBOの中でももっとも難しい問題の１つですので、このテーマを学習することはとても勉強になります。

##15を素因数分解する
例題で15を素因数分解して見ます。
分解する整数をpとqとおくと、

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=15%20%3D%20p*q">

上記のようになります。この式を満たすためのコスト関数は、

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20(15%20-p%20*%20q)%5E2">

となります。pとqの10進数はバイナリ値の量子ビットをつかって下記のように表現ができます。

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=p%20%3D%201%2B2q_0%2B4q_1">

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=q%20%3D%201%2B2q_2">

こちらを代入すると、

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20%5C%7B15-(1%2B2q_0%2B4q_1)(1%2B2q_2)%5C%7D%5E2">

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%2016q_0%5E2q_2%5E2%20%2B%2016q_0%5E2q_2%20%2B%204q_0%5E2%20%2B%2064q_0q_1q_2%5E2%20%2B%2064q_0q_1q_2%20%2B%2016q_0q_1%20%2B%2016q_0q_2%5E2%20-%20104q_0q_2%20-%2056q_0%20%2B%2064q_1%5E2q_2%5E2%20%2B%2064q_1%5E2q_2%20%2B%2016q_1%5E2%20%2B%2032q_1q_2%5E2%20-%20208q_1q_2%20-%20112q_1%20%2B%204q_2%5E2%20-%2056q_2%20%2B%20196">

ここからバイナリの二乗をはずして、

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20128q_0q_1q_2%20%2B%2016q_0q_1%20-%2056q_0q_2%20-%2052q_0%20-%2048q_1q_2%20-%2096q_1%20-%2052q_2%20%2B%20196">

こちらを計算していきます。




##三体問題の分解を使う

ここで、上記に３体問題が出てきますので、QUBOではそのままでは解けませんので２体問題への分解を行います。チュートリアル１８で出てきたテクニックを使い、新しい量子ビットq1*q2 = q3を導入して式に代入し、最後にペナルティ項をつけます。

<img class="math math-inline" src="https://render.githubusercontent.com/render/math?math=E%20%3D%20128q_0q_3%20%2B%2016q_0q_1%20-%2056q_0q_2%20-%2052q_0%20-%2048q_1q_2%20-%2096q_1%20-%2052q_2%20%2B%20196%20%2B%20%5Cdelta(3q_3%2Bq_1q_2-2q_1q_3-2q_2q_3)">

この式を解いていきます。ここで、Gamma = 128として解いて見ると、各項の係数をQUBOmatrixに入れて見て、

In [6]:
import wildqat as wq
a = wq.opt()
a.qubo = [[-52,16,-56,128],[0,-96,80,-256],[0,0,-52,-256],[0,0,0,384]]
a.sa()

1.6109580993652344


[0, 1, 1, 1]

In [7]:
print(a.E[-1])

-196.0


ここで、答えは0,0,0,1がでましたので、
p = 1+0+4=5
q = 1+2 = 3
となりました。

エネルギーも最終のエネルギーが-196となり、上記コスト関数の定数項と足し合わせることで全体のエネルギーが0となっているのが確認できました。

これで素因数分解は解けています。